miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/workflow/DefaultWorkflowManager.java

```diff
@@ -0,0 +1,135 @@
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.Date;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import javax.annotation.Resource;
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+import com.google.common.annotations.VisibleForTesting;
+
+import uk.ac.bbsrc.tgac.miso.core.data.workflow.Progress;
+import uk.ac.bbsrc.tgac.miso.core.data.workflow.ProgressStep;
+import uk.ac.bbsrc.tgac.miso.core.data.workflow.ProgressStep.InputType;
+import uk.ac.bbsrc.tgac.miso.core.data.workflow.Workflow;
+import uk.ac.bbsrc.tgac.miso.core.data.workflow.WorkflowExecutor;
+import uk.ac.bbsrc.tgac.miso.core.data.workflow.impl.ProgressImpl;
+import uk.ac.bbsrc.tgac.miso.core.store.ProgressStore;
+import uk.ac.bbsrc.tgac.miso.service.exception.ValidationError;
+import uk.ac.bbsrc.tgac.miso.service.exception.ValidationException;
+import uk.ac.bbsrc.tgac.miso.service.security.AuthorizationManager;
+import uk.ac.bbsrc.tgac.miso.service.workflow.factory.ProgressStepFactory;
+
+@Service
+@Transactional(rollbackFor = Exception.class)
+public class DefaultWorkflowManager implements WorkflowManager {
+  @Autowired
+  private AuthorizationManager authorizationManager;
+
+  @Autowired
+  private ProgressStore progressStore;
+
+  @Autowired
+  private WorkflowExecutor workflowExecutor;
+
+  @Resource()
+  private Map<FactoryType, ProgressStepFactory> progressStepFactoryMap;
+
+  @Override
+  public Workflow beginWorkflow(String workflowNameString) throws IOException {
+    WorkflowName workflowName = WorkflowName.valueOf(workflowNameString);
+
+    Progress progress = new ProgressImpl();
+    progress.setWorkflowName(workflowName);
+
+    saveProgress(progress);
+    return workflowName.createWorklow(progress);
+  }
+
+  @Override
+  public Workflow processInput(Workflow workflow, String input) throws IOException {
+    workflow.processInput(makeProgressStep(input, workflow.getNextStep().getDataTypes()));
+    saveProgress(workflow.getProgress());
+    return workflow;
+  }
+
+  @Override
+  public Workflow processInput(Workflow workflow, int stepNumber, String input) throws IOException {
+    workflow.processInput(stepNumber, makeProgressStep(input, workflow.getNextStep().getDataTypes()));
+    saveProgress(workflow.getProgress());
+    return workflow;
+  }
+
+  @VisibleForTesting
+  protected ProgressStep makeProgressStep(String input, Set<InputType> inputTypes) throws IOException {
+    for (FactoryType factoryType : getFactoryTypes(inputTypes)) {
+      ProgressStep step = progressStepFactoryMap.get(factoryType).create(input, inputTypes);
+      if (step != null) {
+        return step;
+      }
+    }
+
+    throw new ValidationException(Collections.singletonList(new ValidationError("Could not construct ProgressStep")));
```

Comment:
`ValidationError` messages should be useful to the user, who shouldn't know what a `ProgressStep` is